### PR TITLE
Implement test structure and first preliminary tests

### PR DESCRIPTION
**Pull request by alankbi**
### List of Changes:

* Add integration (API) test and unit test files for existing codebase
* Implement several test cases for SetupService, RunnerService, and other classes
* Create files and basic environment setup for classes that need future test coverage
* Add try finally block to clean up temp folder even if exception occurs

### Notes:

* As of now, the API and generator tests are left blank as I think we need to wait a bit for those to be fleshed out
* RunnerService, SetupService, and some failure tests are implemented (mostly) up to standard (my standards at least lol)
* DockerService is also left blank because it's hard to test (although should be well covered by integration tests by default)

### CODE REVIEW

**Comment by alankbi** [comment]
[Create test structure and preliminary tests for tester service](https://trello.com/c/pPxQDPg9/247-create-test-structure-and-preliminary-tests-for-tester-service)

**Comment by zpChris** [comment]
> some failure tests are implemented (mostly) up to standard (my standards at least lol)

Lots of qualifiers there haha 😄

File: src/test/java/com/rocketden/tester/api/RunnerTests.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    private static final Language LANGUAGE = Language.JAVA;
+
+    @Test
+    public void runRequestBadLanguage() throws Exception {
+        RunRequest request = new RunRequest();
+        request.setCode(CODE);
+        request.setLanguage(LANGUAGE);
+        request.setProblem(new Problem());
+
+        String jsonRequest = "{\"code\": \"print('hi')\", \"language\": \"js\", \"problem\": \"{}\"}";
```
Comment (zpChris) [inline src/test/java/com/rocketden/tester/api/RunnerTests.java:44]: I believe this test fails because there is no language called "js", right -- it's actually called "javascript"? If so, I feel like it'd be better to make this clearer by just making a fake language or a clear typo like "notjava".

File: src/test/java/com/rocketden/tester/api/RunnerTests.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+@AutoConfigureMockMvc
+class RunnerTests {
+
+    @Autowired
+    private MockMvc mockMvc;
+
+    private static final String POST_RUNNER = "/api/v1/runner";
+
+    private static final String CODE = "public class Solution {}";
+    private static final Language LANGUAGE = Language.JAVA;
```
Comment (zpChris) [inline src/test/java/com/rocketden/tester/api/RunnerTests.java:35]: Remove these fields, as these aren't necessary for the test (as it's currently set up).

File: src/test/java/com/rocketden/tester/service/RunnerServiceTests.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        ApiError ERROR = DockerSetupError.BUILD_DOCKER_CONTAINER;
+
+        Mockito.doReturn(TEMP_FOLDER).when(setupService).createTempFolder(request);
+        Mockito.doThrow(new ApiException(ERROR)).when(dockerService).spawnAndRun(Mockito.anyString(), Mockito.any());
+
+        ApiException exception = assertThrows(ApiException.class, () -> runnerService.run(request));
+
+        assertEquals(ERROR, exception.getError());
+        // Despite the exception, the code cleanup step still occurs
+        verify(setupService).deleteTempFolder(TEMP_FOLDER);
```
Comment (zpChris) [inline src/test/java/com/rocketden/tester/service/RunnerServiceTests.java:62]: This was not invoked on when I ran this test locally -- does this test pass for you?

File: src/test/java/com/rocketden/tester/service/generators/JavaDriverGeneratorServiceTests.java [hunk 4]
```diff
@@ -0,0 +1,20 @@
+package com.rocketden.tester.service.generators;
```
Comment (zpChris) [inline src/test/java/com/rocketden/tester/service/generators/JavaDriverGeneratorServiceTests.java:1]: Note: I implement tests in this file, and `PythonDriverGeneratorServiceTests.java`, in #9 -- so I have no preference if you keep or delete these files (either way works great).

File: src/test/java/com/rocketden/tester/util/UtilityTestMethods.java [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+    public static String convertObjectToJsonString(Object o) {
+        Gson gson = new Gson();
+        return gson.toJson(o);
+    }
+
+    public static <T> T toObject(String json, Class<T> c) {
+        Gson gson = new Gson();
+        return gson.fromJson(json, c);
+    }
+}
```
Comment (zpChris) [inline src/test/java/com/rocketden/tester/util/UtilityTestMethods.java:16]: nit: Add ending newline.

File: tester.iml [hunk 6]
```diff
@@ -87,5 +87,7 @@
     <orderEntry type="library" scope="TEST" name="Maven: org.springframework:spring-test:5.2.10.RELEASE" level="project" />
     <orderEntry type="library" scope="TEST" name="Maven: org.xmlunit:xmlunit-core:2.7.0" level="project" />
     <orderEntry type="library" name="Maven: commons-io:commons-io:2.6" level="project" />
+    <orderEntry type="library" name="Maven: commons-lang:commons-lang:2.6" level="project" />
+    <orderEntry type="library" scope="TEST" name="Maven: com.google.code.gson:gson:2.8.6" level="project" />
```
Comment (zpChris) [inline tester.iml:91]: I don't remember having to add these in the main repository -- is this just bad memory or is this for some reason only necessary in `tester.iml`?

**Review by zpChris — CHANGES_REQUESTED** [review]
Requesting changes as some of the tests seem to be failing, and I don't think some of these files are necessary. Otherwise, these tests look great though!

File: src/test/java/com/rocketden/tester/service/DockerServiceTests.java [hunk 7]
```diff
@@ -0,0 +1,20 @@
+package com.rocketden.tester.service;
```
[the remark this answers is not in the document]
Comment (alankbi) [reply src/test/java/com/rocketden/tester/service/DockerServiceTests.java:1]: I'm not mad about skipping these tests at all lol

File: src/test/java/com/rocketden/tester/service/RunnerServiceTests.java (hunk 3 above)
Comment (alankbi) [reply src/test/java/com/rocketden/tester/service/RunnerServiceTests.java:62]: Fixed - I forgot to implement a try-finally block to make the cleanup occur regardless of error (I think you mentioned it in an earlier PR)

File: src/test/java/com/rocketden/tester/service/generators/JavaDriverGeneratorServiceTests.java (hunk 4 above)
Comment (alankbi) [reply src/test/java/com/rocketden/tester/service/generators/JavaDriverGeneratorServiceTests.java:1]: Sounds good - it'll probably get resolved after merging it all

File: tester.iml (hunk 6 above)
Comment (alankbi) [reply tester.iml:91]: Yeah no this is an auto-generated file so this isn't relevant to any functioning code stuff

File: src/test/java/com/rocketden/tester/api/JavaTests.java [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    @Autowired
+    private MockMvc mockMvc;
+
+    private static final String POST_RUNNER = "/api/v1/runner";
+
+    private static final String CODE = "public class Solution {}";
+    private static final Language LANGUAGE = Language.JAVA;
+
+    @Test
+    public void runRequestSuccess() throws Exception {
```
[the remark this answers is not in the document]
Comment (alankbi) [reply src/test/java/com/rocketden/tester/api/JavaTests.java:35]: Response in other thread so I'll resolve this one

File: src/test/java/com/rocketden/tester/api/PythonTests.java [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+	@Autowired
+	private MockMvc mockMvc;
+
+	private static final String POST_RUNNER = "/api/v1/runner";
+
+	private static final String CODE = "print('hello')";
+	private static final Language LANGUAGE = Language.PYTHON;
+
+	@Test
+	public void runRequestSuccess() throws Exception {
```
[the remark this answers is not in the document]
Comment (alankbi) [reply src/test/java/com/rocketden/tester/api/PythonTests.java:35]: Yeah these tests as of now fail - when I wrote them, the driver files were hardcoded and I didn't want to create tests that would only work temporarily. However now that we have the driver code (mostly) working I'll merge the branches and get these ones to pass (although once we figure out the output capturing and flesh out the RunDto return object these tests will have to be updated again)

[the remark this answers is not in the document]
Comment (zpChris) [reply src/test/java/com/rocketden/tester/api/PythonTests.java:35]: Makes sense and sounds good -- if you wanted to just remove that test or leave it blank until that full process is fully implemented I'd be OK with that, but if you want to update them as we update the codebase that's cool too, your choice 👍

**Comment by alankbi** [comment]
Re-requesting review - I pulled the latest changes from `AddProblemDynamic` (also changed this PR to merge into that branch) and since now the driver file stuff mostly works, I implemented several additional tests and error validation checks (so feel free to request changes again cause there's some new stuff)

File: src/main/java/com/rocketden/tester/service/SetupService.java [hunk 10]
```diff
@@ -30,16 +30,16 @@ public SetupService(DriverFileService driverFileService) {
     public String createTempFolder(RunRequest request) {
```
Comment (zpChris) [inline src/main/java/com/rocketden/tester/service/SetupService.java:30]: Could this function throw an error _after_ it creates the temp folder, but perhaps before it makes the driver or solution file? If so, maybe we could add the `deleteTempFolder` function in the inner try-catch block? (I would comment directly there but Github won't let me.)

File: src/main/java/com/rocketden/tester/service/RunnerService.java [hunk 11]
```diff
@@ -18,10 +22,32 @@ public RunnerService(SetupService setupService, DockerService dockerService) {
...
+                throw new ApiException(ProblemError.BAD_PARAMETER_SETTINGS);
+            }
+        });
+
+        String folder = setupService.createTempFolder(request);
+        try {
+            return dockerService.spawnAndRun(folder, request.getLanguage());
+        } finally {
+            // Clean up temp folder regardless if above code throws an exception
+            setupService.deleteTempFolder(folder);
```
Comment (zpChris) [inline src/main/java/com/rocketden/tester/service/RunnerService.java:50]: Thanks for doing this! Good solution ✅

File: src/test/java/com/rocketden/tester/api/JavaTests.java [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.http.MediaType;
+import org.springframework.test.web.servlet.MockMvc;
+import org.springframework.test.web.servlet.MvcResult;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
```
Comment (zpChris) [inline src/test/java/com/rocketden/tester/api/JavaTests.java:22]: nit: Remove unused imports.

File: src/test/java/com/rocketden/tester/api/JavaTests.java [hunk 13]
```diff
@@ -0,0 +1,89 @@
+package com.rocketden.tester.api;
+
+import com.rocketden.tester.model.problem.ProblemIOType;
+import com.rocketden.tester.model.problem.ProblemInput;
+import com.rocketden.tester.model.problem.ProblemTestCase;
```
Comment (zpChris) [inline src/test/java/com/rocketden/tester/api/JavaTests.java:5]: nit: Remove unused imports.

File: src/test/java/com/rocketden/tester/api/JavaTests.java [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+
+    @Autowired
+    private MockMvc mockMvc;
+
+    private static final String POST_RUNNER = "/api/v1/runner";
+
+    private static final Language LANGUAGE = Language.JAVA;
+    private static final String CODE = String.join("\n",
+            "import java.util.Arrays;",
+            "",
```
Comment (zpChris) [inline src/test/java/com/rocketden/tester/api/JavaTests.java:42]: This reminds me that we may want to add `import java.util.*;` to the start of every file for the user's code (on frontend and backend) as well.

File: src/test/java/com/rocketden/tester/api/JavaTests.java [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+            "public class Solution {",
+            "    public int findMax(int[] array) {",
+            "        return Arrays.stream(array).max().getAsInt();",
+            "    }",
+            "}");
+
+    @Test
+    public void runRequestSuccess() throws Exception {
```
Comment (zpChris) [inline src/test/java/com/rocketden/tester/api/JavaTests.java:50]: nit, just an idea: If possible, either in `JavaTests.java` or `PythonTests.java`, could we add tests that verify the Console and Error portions work as well?

I'm not actually sure if this is necessary just because I know this format will change, so I'm 100% ok if we want to go ahead with this test (also because there are 600+ line changes) -- however, I wanted to note it down at least for the future.

File: src/test/java/com/rocketden/tester/api/PythonTests.java [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+class PythonTests {
+
+	@Autowired
+	private MockMvc mockMvc;
+
+	private static final String POST_RUNNER = "/api/v1/runner";
+
+	private static final Language LANGUAGE = Language.PYTHON;
+	private static final String CODE = String.join("\n",
+			"class Solution:",
```
Comment (zpChris) [inline src/test/java/com/rocketden/tester/api/PythonTests.java:34]: This is forward-looking -- similar to Java with `import java.util.*;`, are there any imports we should initially place into the user's file that have commonly-used functions?

File: src/test/java/com/rocketden/tester/service/SetupServiceTests.java [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+
+        String driverFile = String.format("%s/Driver.%s", folder, LANGUAGE.getExtension());
+        String solutionFile = String.format("%s/Solution.%s", folder, LANGUAGE.getExtension());
+
+        verify(driverFileService).writeDriverFile(eq(driverFile), eq(LANGUAGE), eq(problem));
+        assertTrue(new File(solutionFile).exists());
+
+        assertTrue(folder.contains("src/main/java/com/rocketden/tester/temp"));
+
+        // Cleanup any files created (note: this will not always be reached if failures occur)
```
Comment (zpChris) [inline src/test/java/com/rocketden/tester/service/SetupServiceTests.java:56]: This relates to one of my notes above, but maybe we include `deleteTempFolder` in the exceptions as well to prevent this edge case from happening?

File: src/test/java/com/rocketden/tester/service/SetupServiceTests.java [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+
+        assertTrue(new File(folder).exists());
+
+        setupService.deleteTempFolder(folder);
+
+        assertFalse(new File(folder).exists());
+    }
+
+    @Test
+    public void deleteTempFolderInvalidFolder() {
```
Comment (zpChris) [inline src/test/java/com/rocketden/tester/service/SetupServiceTests.java:78]: nit: Perhaps test the `DockerSetupError.DELETE_TEMP_FOLDER` error as well?

File: src/test/java/com/rocketden/tester/util/UtilityTestMethods.java [hunk 19]
```diff
@@ -0,0 +1,16 @@
+package com.rocketden.tester.util;
+
+import com.google.gson.Gson;
+
+public class UtilityTestMethods {
+
+    public static String convertObjectToJsonString(Object o) {
+        Gson gson = new Gson();
```
Comment (zpChris) [inline src/test/java/com/rocketden/tester/util/UtilityTestMethods.java:8]: I know we have the same setup in the main repository, and I may have asked this question before, but could we add the GSON object as a static field in this class?

File: src/test/java/com/rocketden/tester/util/ProblemTestMethods.java [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+import java.util.List;
+import java.util.Map;
+
+public class ProblemTestMethods {
+
+    public static Problem getFindMaxProblem(String... inputs) {
+        Map<Language, String> methodNames = Map.of(
+                Language.JAVA, "findMax",
+                Language.PYTHON, "find_max"
+        );
```
Comment (zpChris) [inline src/test/java/com/rocketden/tester/util/ProblemTestMethods.java:19]: Heads up, per our conversation, you may need to update this if I change the base branch #7.

**Review by zpChris — APPROVED** [review]
LGTM!

I left some comments and nits, but nothing big, and I'm confident you can make the relevant changes (or decide that they're unnecessary).

**Comment by zpChris** [comment]
One more overall nit: There are also some unused imports in `RunnerTests.java`. Do you have a linter that checks for unused imports? No problem if not, just wanted to check ✔️

File: src/main/java/com/rocketden/tester/service/SetupService.java (hunk 10 above)
Comment (alankbi) [reply src/main/java/com/rocketden/tester/service/SetupService.java]: This is referring to if the createTempFolder method fails halfway through right? If so, I split createTempFolder to add an additional method populateTempFolder, which handles the writing of the driver and solution files. This way, createTempFolder can handle just creating and returning the folder name (which we need in order to be able to call delete on the correct location), and then the populating of file can be included in the try catch block so that the finally block will be executed

File: src/test/java/com/rocketden/tester/api/PythonTests.java (hunk 16 above)
Comment (alankbi) [reply src/test/java/com/rocketden/tester/api/PythonTests.java:34]: Good question - I don't use Python for LeetCode-type stuff enough to know if there's anything absolutely critical, but I guess based on that fact I think it should be good? But yeah def important to remember for other languages that might need them

File: src/test/java/com/rocketden/tester/api/JavaTests.java (hunk 15 above)
Comment (alankbi) [reply src/test/java/com/rocketden/tester/api/JavaTests.java:50]: Yeah, I think it'd def be good to have those (though prob implemented later once we implement the proper output parsing RunDto responses). There's a test stub for error occurred but I'll add ones too for testing console output

File: src/test/java/com/rocketden/tester/service/SetupServiceTests.java (hunk 17 above)
Comment (alankbi) [reply src/test/java/com/rocketden/tester/service/SetupServiceTests.java:57]: I'm not sure if I totally understand the exceptions part, but I did move this code into a try finally block so the deleteTempFolder should now always execute (same thing for the main codebase as we talked about in different comments)

Lmk if you had something else in mind and I can implement that as well (tho maybe in the next PR if this one gets merged before then haha)

File: src/test/java/com/rocketden/tester/util/ProblemTestMethods.java (hunk 20 above)
Comment (alankbi) [reply src/test/java/com/rocketden/tester/util/ProblemTestMethods.java:19]: Sounds good, I'll merge main into this branch (it doesn't show any warnings/merge conflicts on GitHub which is concerning given that I immediately get a compilation error locally after merging haha)

File: src/test/java/com/rocketden/tester/service/SetupServiceTests.java (hunk 18 above)
Comment (alankbi) [reply src/test/java/com/rocketden/tester/service/SetupServiceTests.java:81]: I haven't been able to get that code to really fail based on any of our use cases - it requires us to catch the IOException which is why I added the ApiError, but imo given that it's a production-ready package I feel like we might be able to get away with not testing that

I wasn't overly rigorous on playing around with it tho so if in the future it might be useful then I'd def be down on adding some kind of test for that

**Comment by alankbi** [comment]
> One more overall nit: There are also some unused imports in `RunnerTests.java`. Do you have a linter that checks for unused imports? No problem if not, just wanted to check ✔️

Removed them - I do have linting haha although I guess I'm prone to miss them at times

Also, there were another few decent rounds of changes made - I'll merge this for now, but if you notice anything feel free to comment and I can address those in future PRs
